Add docs for OTLP meter registry. (#308)
micrometer-metrics/micrometer-docs: src/docs/implementations/otlp.adoc

[%autowidth]
|===
|Key | Default value

|telemetry.sdk.name
|io.micrometer

|telemetry.sdk.language
|java

|telemetry.sdk.version
|<micrometer-core-version> (E.g: 1.11.0)

|service.name
|unknown_service
|===

If this config is empty, then resource attributes will be loaded from the environmental variable `OTEL_RESOURCE_ATTRIBUTES`. `service.name` can be overridden by the environmental variable `OTEL_SERVICE_NAME` and this takes precedence over other configs.

== Supported metrics
https://opentelemetry.io/docs/specs/otel/metrics/data-model/#metric-points[Metric points] define the different data-points that are supported in OTLP. Micrometer supports exporting the below data-points in OTLP format,

1. https://opentelemetry.io/docs/specs/otel/metrics/data-model/#sums[Sums]
2. https://opentelemetry.io/docs/specs/otel/metrics/data-model/#gauge[Gauge]
3. https://opentelemetry.io/docs/specs/otel/metrics/data-model/#histogram[Histogram]
4. https://opentelemetry.io/docs/specs/otel/metrics/data-model/#summary-legacy[Summary]

The below table maps OTLP datapoint and the micrometer meters,
[%autowidth]
|===

> |OTLP DataPoint | Micrometer meter type

Reviewer: I'm thinking if it would make more sense to do the mapping the other way Micrometer -> OTLP and also add Meter, TimeGauge, MultiGauge.

Reviewer: @lenin-jaganathan GH somewhat hides this but what do you think?

Author: I initially started with that approach and I saw that table grew big, so I kind of moved away from that. Let me re-visit that. Also, will it be fine if we do N-1 mapping for micrometer meters or do you think 1-1 is needed?

Reviewer: N-1 is completely good, please feel free to keep it as is if it makes more sense to you, but could you please add `Meter`, `TimeGauge`, `MultiGauge` too? I guess all of them translate to gauge in OTLP.

Author: In the case of Otlp, I guess MultiGauge is not implemented (same as an arbitrary Meter). Also, currently, we are not exporting anything for the `Meter`. Do you agree with the 2 statements?

I haven't verified the behaviour but by looking at the code, it seems so.

Reviewer 2: MultiGauge is a bit of a misnomer in that it itself is not a Meter, but you use its API to register gauges.
Meter will I think publish the metadata about the meter but without any datapoints. We still need to implement that.

|Counter, FunctionCounter
| Sums

|Gauge
|Gauge

|Histogram
|Timer, DistributionSummary, LongTaskTimer, FunctionTimer (only sum and count are set)

|Summary
|Timer, DistributionSummary, LongTaskTimer
|===

> *Note*: `max` on Histogram datapoint is only supported in Delta Aggregation Temporality.

Reviewer: Could you please add a sentence why it is not supported on cumulative?
(Like OTLP defines it as max from app startup which is not very useful and Micrometer has a time window max).

Author: Added.

== Histograms and Percentiles

> Micrometer `Timer` and `DistributionSummary` supports configuring http://localhost:3000/docs/concepts#_histograms_and_percentiles[client side percentiles and percentile histogram]. OTLP specification terms Summary Data-point (client-side percentiles) as legacy and not recommended for new applications. Summary data point also cannot have min/max associated with it. Due these reasons micrometer prefers exporting Timers and DistributionSummary as Histogram datapoint. By default a Timer/DistributionSummary without any additional percentile/histogram config will be exported as Histogram Datapoint. However, by configuring the timer to generate only client side percentiles using `publishPercentiles` this can be changed to a Summary data point exporting pre-calculated percentiles. When both `publishPercentiles` and (`publishPercentileHistogram` or `serviceLevelObjectives`) are configured Histogram Datapoint is preferred and pre-calculated percentiles *will not* be generated. See the below table on which data point will be used with different configurations,

Reviewer: - `https://micrometer.io` instead of `http://localhost:3000` but I think the link dhould be ok if you delete the schema+host+port part
- histogram -> histograms

```suggestion
Micrometer `Timer` and `DistributionSummary` supports configuring docs/concepts#_histograms_and_percentiles[client side percentiles and percentile histograms]. OTLP specification terms Summary Data-point (client-side percentiles) as legacy and not recommended for new applications. Summary data point also cannot have min/max associated with it. Due these reasons micrometer prefers exporting Timers and DistributionSummary as Histogram datapoint. By default a Timer/DistributionSummary without any additional percentile/histogram config will be exported as Histogram Datapoint. However, by configuring the timer to generate only client side percentiles using `publishPercentiles` this can be changed to a Summary data point exporting pre-calculated percentiles. When both `publishPercentiles` and (`publishPercentileHistogram` or `serviceLevelObjectives`) are configured Histogram Datapoint is preferred and pre-calculated percentiles *will not* be generated. See the below table on which data point will be used with different configurations:
```

[%autowidth]
|===
|Configuration | OTLP Data point

| publishPercentiles
| Summary

| publishPercentileHistogram
| Histogram

| serviceLevelObjectives
| Histogram

| publishPercentiles and (publishPercentileHistogram/serviceLevelObjectives)
| Histogram
|===

> Alternatively, If you are using spring boot you can use the https://docs.spring.io/spring-boot/docs/3.1.0/reference/htmlsingle/#actuator.metrics.customizing.per-meter-properties[per meter properties] to configure this behaviour.

Reviewer: I'm thinking if we should use `current` instead of `3.1.0`.
```suggestion
Alternatively, If you are using Spring Boot you can use the https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#actuator.metrics.customizing.per-meter-properties[per meter properties] to configure this behaviour.
```

> If you want to generate Histogram Data point for a timer with name `test.timer` with default buckets generated by micrometer use,

Reviewer: ```suggestion
If you want to generate Histogram Data point for a Timer with name `test.timer` with default buckets generated by micrometer use:
```

[source.properties]
----
management.metrics.distribution.percentiles-histogram.test.timer=true
----

> and for buckets with customized slo,

Reviewer: ```suggestion
and for buckets with customized slo:
```

[source.properties]
----
management.metrics.distribution.slo.test.timer=10.0,100.0,500.0,1000.0
----

> Alternatively, if you want to generate Summary Data point for a timer with name `test.timer` with 90th and 99th percentiles you can use the below config,

Reviewer: ```suggestion
Alternatively, if you want to generate Summary Data point for a timer with name `test.timer` with 90th and 99th percentiles you can use the below config:
```
